Support fixme's in docstrings

badsketch opened the pull request:
<!--
Thank you for submitting a PR to pylint!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

- [ ] Document your change, if it is a non-trivial one.
  - A maintainer might label the issue ``skip-news`` if the change does not need to be in the changelog.
  - Otherwise, create a news fragment with ``towncrier create <IssueNumber>.<type>`` which will be
    included in the changelog. ``<type>`` can be one of the types defined in `./towncrier.toml`.
    If necessary you can write details or offer examples on how the new change is supposed to work.
  - Generating the doc is done with ``tox -e docs``
- [ ] Relate your change to an issue in the tracker if such an issue exists (Refs #1234, Closes #1234)
- [ ] Write comprehensive commit messages and/or a good description of what the PR does.
- [ ] Keep the change small, separate the consensual changes from the opinionated one.
  Don't hesitate to open multiple PRs if the change requires it. If your review is so
  big it requires to actually plan and allocate time to review, it's more likely
  that it's going to go stale.
- [ ] If you used multiple emails or multiple names when contributing, add your mails
      and preferred name in ``script/.contributors_aliases.json``
-->

## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
|    | :bug: Bug fix          |
| ✓   | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Description

<!-- If this PR references an issue without fixing it: -->


<!-- If this PR fixes an issue, use the following to automatically close when we merge: -->

Closes #9255 


Previous PR discussion here: https://github.com/pylint-dev/pylint/pull/9281
- now an enhancement of existing `fixme` rather than a new message
- `check-fixme-in-docstring` is the setting that enables it and defaults to false
- also suggestions to improve the existing description: https://github.com/pylint-dev/pylint/pull/9281#discussion_r1415102708. I kind of like how the message is the TODO itself. As for the "Used when a warning note..." is that for a tooltip? I believe almost all checker messages have something like this, right?

Appreciate any feedback!

badsketch commented on pylint/checkers/misc.py:
```diff
@@ -133,16 +148,29 @@ def process_tokens(self, tokens: list[tokenize.TokenInfo]) -> None:
+            elif self.linter.config.check_fixme_in_docstring and self._is_docstring_comment(token_info):
+                docstring_lines = token_info.string.split("\n")
+                for line_no, line in enumerate(docstring_lines):
+                    comment_text = line.removeprefix('"""').lstrip().removesuffix('"""')  # trim '""""' and whitespace
```
`removeprefix()` and `removesuffix()` are new in python 3.9. Dumb question, but how do I tell what version this project supports?

badsketch replied:
After seeing 3.8 test suites fail, I'm guessing I need to make this compatible 😅

DanielNoord replied:
Do we really need this? Can't we just put the full docstring in the message for now?

badsketch replied:
For a case like 
```
"""
TODO msg1
TODO msg2
"""
```
this PR will create two TODO lint messages. If we put the full docstring for both, it might be confusing or overly wordy, no?

badsketch commented:
When running `pylint -h`, I get 
```
Miscellaneous:
  BaseChecker for encoding issues.

  --notes <comma separated values>
                        List of note tags to take in consideration, separated by a comma. (default:
                        ('FIXME', 'XXX', 'TODO'))
  --notes-rgx <regexp>  Regular expression of note tags to take in consideration. (default: )
  --check-fixme-in-docstring <y or n>
                        Whether or not to search for fixme's in docstrings. (default: False)
```
Thoughts on updating the docstring to be "Checker for encoding issues and fixme notes"? instead of "BaseChecker for encoding issues."

DanielNoord commented:
> When running `pylint -h`, I get
> 
> ```
> Miscellaneous:
>   BaseChecker for encoding issues.
> 
>   --notes <comma separated values>
>                         List of note tags to take in consideration, separated by a comma. (default:
>                         ('FIXME', 'XXX', 'TODO'))
>   --notes-rgx <regexp>  Regular expression of note tags to take in consideration. (default: )
>   --check-fixme-in-docstring <y or n>
>                         Whether or not to search for fixme's in docstrings. (default: False)
> ```
> 
> Thoughts on updating the docstring to be "Checker for encoding issues and fixme notes"? instead of "BaseChecker for encoding issues."

Fine with me!

DanielNoord commented on pylint/checkers/misc.py:
```diff
@@ -133,16 +152,41 @@ def process_tokens(self, tokens: list[tokenize.TokenInfo]) -> None:
+    def _is_docstring_comment(self, token_info: tokenize.TokenInfo) -> bool:
+        return (
+            token_info.type == tokenize.STRING
+            and token_info.line.lstrip().startswith('"""')
```
Note that a docstring can also start with `'''`. I'm wondering if this should live in this tokeniser checker as I think it is actually quite hard to recognise docstrings on tokens alone.

Have you considered doing it as a checker for `nodes.Module`, `nodes.ClassDef`, etc? Then you can just check if the regex is in the `.doc` attribute.

badsketch replied:
>Note that a docstring can also start with '''

 totally forgot about this, thanks!

>Have you considered doing it as a checker for nodes.Module, nodes.ClassDef, etc? Then you can just check if the regex is in the .doc attribute.

 I was considering the possibility docstrings could appear outside of `nodes.Module` and `nodes.classDef`, so I figured it's best to use the existing token stream to watch for all occurrences. However, you could argue it's not good python practice (?) in the first place to have docstrings outside modules/classes/methods. Agreed the PR doesn't use the safest heuristic to determine if it's a docstring.

So it looks like we could do
1. update `_is_docstring_comment()` to also check `startswith("'''")`
2. refactor to use `nodes.Module`, `nodes.ClassDef`, `nodes.FunctionDef` and we tighten the scope of docstring fixmes
3. update tokenizer with a new docstring token similar to how we have a token.COMMENT type for a comment fixme. (Haven't looked too deep into this, will probably be higher effort)

Totally down to change it to 2, but would users claim false negatives when they try to create docstring fixme's outside of module/classes/function defs? Perhaps it would help if there were a lint message that recommends against docstrings outside of those places. Do we have something like that already?

DanielNoord replied:
@Pierre-Sassoulas Opinion? I think trying option 1 for now might be fine, we can always refactor to 2 later on. I just thought I would raise the question to see if it was consciously ignored.

badsketch replied:
Yeah we already use tokenizer for comments. If I understand correctly, tokenizer is for instances where there's not a defined node for the check (eg, a comment can appear "anywhere" in the code, so we check for all tokens for that appearance). I tried to follow that logic with docstrings. Hence I piggy back on the tokenizer to examine for occurrences of `"""` and `'''`.  If we decide we only wish to support docstrings in classes, functions, and methods, then I could use a node visitor on those 3 node types.

It may be easiest at this point to go with Option 1 at this point since it'd be straightforward to make the change in the PR. And as Daniel mentioned, we could refactor to 2 in the future. Thoughts?

Pierre-Sassoulas replied:
Docstrings are not supposed to be everywhere in a module, that's useless statements otherwise, so a node approach would work. But if comments requires tokenizer, for consistency of aproachs let's go with 1)

DanielNoord commented on tests/checkers/unittest_misc.py:
```diff
@@ -116,3 +117,151 @@ def test_dont_trigger_on_todoist(self) -> None:
         with self.assertNoMessages():
             self.checker.process_tokens(_tokenize_str(code))
+
+    @set_config(check_fixme_in_docstring=True)
+    def test_docstring_with_message(self) -> None:
```
@Pierre-Sassoulas Do you want these unittests? I would be fine with having only the functional tests and remove these. They feel like duplicates

badsketch replied:
Gotcha, that makes sense! Reverted changes to this file.

DanielNoord commented on pylint/checkers/misc.py:
```diff
@@ -90,18 +90,35 @@ class EncodingChecker(BaseTokenChecker, BaseRawFileChecker):
+            self._comment_fixme_pattern = re.compile(comment_regex, re.I)
+            if self.linter.config.check_fixme_in_docstring:
+                docstring_regex = rf"((\"\"\")|(\'\'\'))\s*({notes})(?=(:|\s|\Z))"
+                self._docstring_fixme_pattern = re.compile(docstring_regex, re.I)
```
I think we need to always set these two pattern attributes so that we don't get an `AttributeError` in unexpected ways. In the old version we also did this.

badsketch replied:
Gotcha, I've removed the conditionals here and just let `process_tokens()` handle the logic of what to parse if check docstring for fixme setting is enabled.

DanielNoord commented on pylint/checkers/misc.py:
```diff
@@ -133,16 +150,43 @@ def process_tokens(self, tokens: list[tokenize.TokenInfo]) -> None:
+                docstring_lines = token_info.string.split("\n")
+                for line_no, line in enumerate(docstring_lines):
+                    # trim '"""' at beginning or end and whitespace
+                    if line.startswith(('"""', "'''")):
+                        line = line[3:]
+                    line = line.lstrip()
+                    if line.endswith(('"""', "'''")):
+                        line = line[:-3]
+                    if self._docstring_fixme_pattern.search(
+                        '"""' + line.lower()
+                    ) or self._docstring_fixme_pattern.search("'''" + line.lower()):
```
Now that you have fixed the pattern, is this really necessary? I would prefer a pattern where we don't have to do a lot of complicated changes to the lines itself.

badsketch replied:
True, it was getting pretty messy. I've decided to rework the logic to use regex capture groups for both docstrings **and** comments (since in that case we were also sort of changing the line then stitching it together to examine with regex).

DanielNoord reviewed (commented):
Awesome you got this to work!

badsketch commented:
Apologies for the glacial pace of this PR 😬, really appreciate all the feedback! Some updates:
- previously we were taking the token line and using string manipulation to get the fixme message, then rebuilding it so the regex could find it. I've since changed the logic to use just regex with capture groups instead. I've made this change for both docstrings and comments
- To do this, I also had to split the docstring regex pattern between single and multiline. I tried it with just a single regex, but it got really ugly and I think the logic is much cleaner/intuitive if there's two. There's 3 regex patterns total now.

Let me know your thoughts!

badsketch commented:
Those primer tests are awesome! They thankfully caught something I missed when refactoring the comment regex pattern. I was no longer supporting cases like:
```
#     # TODO: something something
```
which is kind of curious as I wonder if that should've been supported in the first place? Regardless, I've updated it so it's supported and added that as an additional functional test.

badsketch commented:
In my last few messages, I mentioned updating how the message was extracted in comment-based fixme's so we wouldn't have to use string manipulation(https://github.com/pylint-dev/pylint/pull/9744/commits/198fd93ca02ba16b9619d2a42d218ad09b5cfa6d, https://github.com/pylint-dev/pylint/pull/9744/commits/1ce441c7960421ddd36fb71ea7c346e5652da978). I've decided to revert that change because it was causing a lot of primer test failures. I think this is due to how comment-based fixme's might be a little inconsistent in its current state:
```
#   TODO: msg1                                        
#   # TODO: msg2
# something # TODO: msg3
# something TODO: msg4
```
results in:
```
test.py:1:1: W0511: TODO: msg1 (fixme)
test.py:2:1: W0511: # TODO: msg2 (fixme)
test.py:3:1: W0511: something # TODO: msg3 (fixme)
```
I wasn't able to replicate what we're doing today using regex capture groups so in favor of not causing any disruptions, I decided to go back to the current method.

If we're open to standardizing some of the decided behavior maybe in the future and allowing changes to the primer tests, I'm also down to discuss!

DanielNoord commented:
@badsketch With the danger of introducing more back and forth: would you be willing to create a proposal for this standardization and apply it to this PR? We can then see the test output and determine whether we are okay with that. That makes it a lot easier to discuss such standardization (with having some examples).

badsketch commented:
@DanielNoord Sure thing, is there a formal way of making these proposals for Pylint features and do I submit it somewhere? Or do you mean try to consider all the usage out there and come back with a more detailed comment?

DanielNoord commented:
No, I meant: just write the code as you would want and make the tests pass. If the test output is acceptable I would be okay with accepting your proposed code.

I don't think we need a full proposal, just the code as you would propose to write it and then being able to see what changes that would give to the test output.

badsketch commented:
@DanielNoord 
Gotcha, I've updated the functional tests for both comment and docstring fixme's to serve as my final proposal. The latest primer test results also align with what I've proposed. They boil down to:
- comment fixme's must start with a `#` then any number of spaces (and only spaces), the fixme keyword, followed by a message
```
# TODO valid
#              TODO valid
# invalid TODO msg
```
- single line docstring fixme's must start with 3 single/double quotes, then any number of spaces (and only spaces), the fixme keyword, followed by a message
```
'''TODO valid'''
""" TODO valid """
''' invalid TODO msg """
```
- multi-line docstring fixme's are within a docstring block, but must be at the beginning of the line. Any number of indentations before it are okay.
```
'''
TODO valid
invalid TODO msg
'''
def foo():
      """
      TODO valid
      """
```

The majority of changed primer tests are for comment fixme's that no longer emit. They appear to be fixme's that are part of a **larger section of code** that was commented out, suggesting the fixme is no longer valid. I think it makes sense for those to no longer be emitted.

DanielNoord reviewed (approved):
LGTM!

@Pierre-Sassoulas Do you want to give this a review as well?

Pierre-Sassoulas commented on doc/whatsnew/fragments/9255.feature:
```diff
@@ -0,0 +1,3 @@
+Add ability for `fixme` check to also search through docstrings.
```
```suggestion
The `fixme` check can now search through docstrings as well as comments, by using ``check-fixme-in-docstring = true`` in the ``[tool.pylint.miscellaneous]`` section.
```

Pierre-Sassoulas reviewed (commented):
Thank you for working on this @badsketch !

I'm wondering why the regex contains `` rf"((\"\"\")|(\'\'\'))``, but we also do `` (token_info.line.lstrip().startswith(('"""', "'''")))`` somewhere else ? My intuition would be that we could just search for the user defined "TODO" regex in comments and docstrings ? Is this because we want the option to not raise for docstring (for semver/compat with the old behavior) so we need to distinguish the two internally ? I didn't make any benchmarks for this but I think this kind of regex executed on each token can have huge performance impact.

badsketch commented:
@Pierre-Sassoulas 
That's a fair point. I think I can optimize it a little by making the`((\"\"\")|(\'\'\'))` non-capture groups. Would that suffice?

>My intuition would be that we could just search for the user defined "TODO" regex in comments and docstrings ? Is this because we want the option to not raise for docstring (for semver/compat with the old behavior) so we need to distinguish the two internally ?

Hm, I'm little unclear what you mean. We're using 3 regexes (comment, single-line docstring, multi-line docstring) for clarity since it's hard to capture all of it in one regex. Also, so that, yes, we don't need to execute a more expensive docstring regex if the option is switched off. Not sure if I answered your question :/

To your point above, we could check if a TODO is present so we don't bother checking for a regex if that keyword isn't there:
```diff
    def _is_multiline_docstring(self, token_info: tokenize.TokenInfo) -> bool:
        return (
            token_info.type == tokenize.STRING
            and (token_info.line.lstrip().startswith(('"""', "'''")))
+           and 'TODO' in token_info.line
            and "\n" in token_info.line.rstrip()
        )
```
What might complicate this is `note-rgx` is enabled, then we'd be executing a regex anyways 
```diff
    def _is_multiline_docstring(self, token_info: tokenize.TokenInfo) -> bool:
        return (
            token_info.type == tokenize.STRING
            and (token_info.line.lstrip().startswith(('"""', "'''")))
+           and 'TODO' in token_info.line.search(TODO|XXX|FIXME|{notes.rgx|)
            and "\n" in token_info.line.rstrip()
        )
```

> I didn't make any benchmarks for this but I think this kind of regex executed on each token can have huge performance impact.

I'd be willing to run a profile to make a more convincing argument. Would following this guide https://pylint.readthedocs.io/en/stable/development_guide/contributor_guide/profiling.html be what you had in mind?

Pierre-Sassoulas commented:
Sorry for the delay, busy week.

> We're using 3 regexes (comment, single-line docstring, multi-line docstring) for clarity since it's hard to capture all of it in one regex.

Ok I understand now. 

If we consider that fixme not being detected in docstring is a false negative that we're fixing now, we can remove the new ``check-fixme-in-docstring`` option. I don't see who would want to track TODO in comments but not in docstring so forcing user to activate the option make pylint worse. What are your thoughts on the matter @DanielNoord ?

I was wondering if the distinction between the 3 regexes, can be removed if we remove the ``check-fixme-in-docstring`` option (and only search for ``notes`` after detecting that we have a comment or a docstring without making any distinction). Tbf the less regex we have to maintain the happier I am (especially since we need to consider thing like catastrophic backtracking that can bring CVE and are hard to catch with blind benchmarks).

> Would following this guide https://pylint.readthedocs.io/en/stable/development_guide/contributor_guide/profiling.html be what you had in mind?

Yes.

DanielNoord commented:
I think we previously decided that we didn't want to consider the docstring case a false negative to make the change not too disruptive for exciting comments. I already approved merging this into the `fixme` warning, I would be okay with going further and not even creating the option but that gives more work to @badsketch again..

Pierre-Sassoulas commented:
Right, how about taking a decision based on the benchmark results ? (If making a distinction is slower we remove the distinction)

badsketch commented:
Please let me know if the following is an accurate profiling case. I ran it on the pylint base and measured cumulative time of `process_tokens()` in `misc.py`

`main` without `check-fixme-in-docstring`
```
ncalls  tottime  percall  cumtime  percall filename:lineno(function)
175    0.011    0.000    0.013    0.000 /Users/moss/projects/pylint/pylint/checkers/misc.py:131(process_tokens)
```

Previously this PR with `check-fixme-in-docstring` on
```
ncalls  tottime  percall  cumtime  percall filename:lineno(function)
175    0.073    0.000    0.120    0.001 /Users/moss/projects/pylint/pylint/checkers/misc.py:150(process_tokens)
```

Considering total `cumtime` has almost 10x, I assume this isn't insignificant.

**BUT** I just made a change in the most recent commit (https://github.com/pylint-dev/pylint/pull/9744/commits/d06c5622a9b9ffa6f977807c827b2c1eb027f551) and was able to shave it down to
```
ncalls  tottime  percall  cumtime  percall filename:lineno(function)
175    0.048    0.000    0.077    0.000 /Users/moss/projects/pylint/pylint/checkers/misc.py:150(process_tokens)
```

I think this is a bit more reasonable since cumulative percall is back down to 0.000?

> I was wondering if the distinction between the 3 regexes, can be removed if we remove the check-fixme-in-docstring option (and only search for notes after detecting that we have a comment or a docstring without making any distinction). Tbf the less regex we have to maintain the happier I am (especially since we need to consider thing like catastrophic backtracking that can bring CVE and are hard to catch with blind benchmarks).

Forgive me if I'm still misunderstanding, but I'm not sure if removing the distinction would help. I tried using one regex originally for both comments and docstrings, but having a one-size fits all regex becomes unwieldy. Most of the complexity arises when trying to extract the TODO message to be outputted. The multi-line docstrings have some edge scenarios when there are multiple on different lines with different spacings. 

```
TokenInfo(type=3 (STRING), string='""" TODO: do this """', start=(2, 0), end=(2, 21), line='""" TODO: do this """\n')
TokenInfo(type=3 (STRING), string='"""\n        TODO: do this\n        TODO: do that\n"""', start=(3, 0), end=(6, 3), line='"""\n        TODO: do this\n        TODO: do that\n"""')

```
I'm worried I'd produce a regex that's harder to understand with greater risk of potential catastrophic backtracking.

jacobtylerwalls commented:
> Could be false negatives, will look into this.

The primer tests still suffer from a fluctuating baseline, so you'll see unrelated messages moving in and out from time to time. We haven't yet identified the source of the indeterminacy. Surely it's a lack of a fixed ordering somewhere. Don't worry about it.

jacobtylerwalls commented:
FYI @badsketch, thanks to some ace detective work by @akamat10, the primer is more stable now, so if you merge main you should see a more informative primer diff.

badsketch commented:
@jacobtylerwalls That's fantastic, appreciate the update! I merged from main and pushed just now.

As a refresher, since it's been a while, the performance profiling results are in this comment: https://github.com/pylint-dev/pylint/pull/9744#issuecomment-2249498558

Also, the latest primer diff above makes sense to me. Most of the instances are where a TODO is part of a section that has been commented out. It would make sense for those to no longer alert as a fixme. The other instances are where a TODO is behind another comment like "# some note # TODO: xyz", which I would also like to propose as an acceptable primer test output change for the sake of consistency. Let me know if there are any issues 👍

badsketch commented:
@jacobtylerwalls yeah for sure. Since there's already a https://github.com/pylint-dev/pylint/blob/main/doc/whatsnew/fragments/9255.feature, do I rename that to 9255.breaking and modify it, or do I create a new fragment?

jacobtylerwalls commented:
I think I like having two separate fragments for this.
